test_move scripts (#437)
UniversalRobots/Universal_Robots_ROS_Driver: ur_robot_driver/doc/usage_example.md

# Using the ur_robot_driver

This tutorial should help you getting started to use the `ur_robot_driver` in your application. It

> explains how to start the robot driver and how to control the robot either using MoveIt! or a simple

Reviewer: ```suggestion
explains how to start the robot driver and how to control the robot using either MoveIt! or a simple
```

on how to do that.

### Real robot
If you have a real robot and want to use it for testing, make sure it is connected to the ROS
machine on which you'd like to run the driver. We recommend a direct connection between the robot
and the control PC so that other network participants will not disturb the network traffic. Make
sure that the network is configured, such that both, the robot and the ROS machine are in the same
subnet. See the robot's user manual how to setup a static IP address on the robot.

### Using a simulated robot
For simulation we recommend running URSim in a virtual machine. You can directly download a virtual
machine image containing a preconfigured URSim from Universal Robot's [download
page](https://www.universal-robots.com/download/?filters[]=98759&query=). Please note, that you will
have to select "**Non-Linux Offline simulator**" in order to get the virtual machine.

The following will assume that you have VirtualBox installed on your machine. If not, you can
install it by calling `sudo apt install virtualbox` in a terminal.

Once downloaded, extract the archive and open the file `URSim_VIRTUAL-<version>.vbox`. It should

> right open in VirtualBox,

Reviewer: ```suggestion
right open in VirtualBox.
```

#### Network setup
Next, setup a host network that will be used for the machine.

For this, click **File** > **Host Network Manager**. A dialog will open in which you can click the
**create** button which will add an entry to the list. It should automatically get the static IP
address `192.168.56.1`. If not, set it up manually in the dialog. Once done, you can close the
dialog again to go back to your VirtualBox main window. There, select the URSim_VIRTUAL... machine
and click on **Settings**. Select the **Network** point in the list on the left and activate at
least one network adapter, setting the **Attached to** dropdown menu to **Host-only Adapter**.

> Select the network you just created, if there are multiple host-only networks available. Click on

Reviewer: ```suggestion
If there are multiple host-only networks available, select the network you just created. Click on
```

the **ok** button to close the dialog and go back to your main window.

As a side note: A host-only network is not the only option to get this running, but it is the
easiest. If you consider using another networking mode, make sure that the virtual machine is

> reachable via its on IP address from the host machine and that it can reach the host machine, as

Reviewer: ```suggestion
reachable via its on IP address from the host machine and that it can reach the host machine as
```

well.

#### Starting a robot

> To start a simulated robot, select your URSim virtual machine and clock on the **Start button**. The

Reviewer: ```suggestion
To start a simulated robot, select your URSim virtual machine and click on the **Start button**. The
```

URCap as explained
[here](https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/blob/master/README.md#setting-up-a-ur-robot-for-ur_robot_driver)
or by using the [headless
mode](https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/blob/master/ur_robot_driver/README.md#headless-mode).
We assume the use of the URCap in the following.

Before any control of the robot can happen, a program containing the **External Control** program
node must be running on the robot. You should see the output `Robot ready to receive control
commands.` in the terminal where you launched the driver. See [this troubleshooting
entry](https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/blob/master/README.md#i-started-everything-but-i-cannot-control-the-robot)
if you experience problems in this step.

### Control the robot using the `test_move` script
Once you've setup the robot for getting controlled by ROS, you can move it around using the
[`test_move`](https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/blob/master/ur_robot_driver/scripts/test_move) script.

Start the driver and an RViz window as before. Then, in a third terminal run

```

> rosrun ur_robot_driver tes_move

Reviewer: ```suggestion
rosrun ur_robot_driver test_move
```

about the different controllers, please see [TODO: Change once merged #423](https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/pull/423).

It will first ask about the controller to be used:

```
Available trajectory controllers:
0 (joint-based): scaled_pos_joint_traj_controller
1 (joint-based): scaled_vel_joint_traj_controller
2 (joint-based): pos_joint_traj_controller
3 (joint-based): vel_joint_traj_controller
4 (joint-based): forward_joint_traj_controller
5 (Cartesian): pose_based_cartesian_traj_controller
6 (Cartesian): joint_based_cartesian_traj_controller
7 (Cartesian): forward_cartesian_traj_controller
Please choose a controller by entering its number (Enter '0' if you are unsure / don't care): 0
```

After that, you'll have to confirm that the motions the robot is about to make, can safely be
executed the way your robot is mounted. If you're running a simulated robot, you can confirm this
without thinking twice. However, if a real robot is connected, please make sure that the robot is
actually free of any obstacles inside the space where the motion is taking place. See the images
below for a visualization of the setpoints:

> **Joint-based motions**: (The cable leaving the base to the back on an e-Series robot)

Reviewer: When looking at this in the rendered .md file, it seems a little unorganized (one on top and two below). Might also be my current screen width interfering with this. As an alternative, I would propose using a single image for this (and also the Cartesian illustrations). This could be done with either using [RViz' trail mechanism](https://answers.ros.org/question/199552/rviz-robotmodel-base_link-show-trail-history/), or using an overlay of partly transparent images. If the screenshots are taken from the same camera angle, the second should be possible.

Author: I've created an overlayed image of all: 
![image](https://user-images.githubusercontent.com/491645/130206477-ec84205e-86f4-4491-8116-8748996b6369.png)

It looks a bit confusing, but it very clearly shows the configurations being used inside the trajectory, which is what we want in that case. @stefanscherzinger do you think, this might be the way to go?

Reviewer: A little dark, but I like it! Let's use that.

Author: Resolved in 426ea83

![joint_based pose 1](resources/usage_example/joint_based_motion_1.png "Pose one of joint-based motion")
![joint_based pose 2](resources/usage_example/joint_based_motion_2.png "Pose two of joint-based motion")
![joint_based pose 3](resources/usage_example/joint_based_motion_3.png "Pose three of joint-based motion")

**Cartesian motions**: (The cable leaving the base to the back on an e-Series robot)
**Please note: Depending on your starting configuration there might also be a valid elbow-down
configuration for each of the poses!**

> ![cartesian pose 1](resources/usage_example/cartesian_motion_1.png "Pose one of Cartesian motion")

Reviewer: See comment above about a single image for this.

Author: Resolved in 426ea83

![cartesian pose 2](resources/usage_example/cartesian_motion_2.png "Pose two of Cartesian motion")
![cartesian pose 3](resources/usage_example/cartesian_motion_3.png "Pose three of Cartesian motion")
![cartesian pose 4](resources/usage_example/cartesian_motion_4.png "Pose four of Cartesian motion")
![cartesian pose 5](resources/usage_example/cartesian_motion_5.png "Pose five of Cartesian motion")

You can play around by changing the trajectories inside the python code, if you like. You can also
setup velocities in each waypoint to get a smooth motion through all waypoints.

### Control the robot using MoveIt!

> For controlling the robot using MoveIt! start the following three launch files (Each in one terminal

Reviewer: ```suggestion
For controlling the robot using MoveIt! start the following three launch files (Each in a separate terminal
```
